File: pset6/credit.py
from cs50 import get_int

cardNum = 0
while (cardNum < 4):
    cardNum = get_int("Number: ")

print(f"Your number: {cardNum}")
digitsList = [34, 37, 50, 51, 52, 53, 54, 55]
list4 = [39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]

length = 0
oddSum = 0
evenSum = 0
valid = True
last2Digits = 0
lastDigit = 0
while (cardNum):
    digit = cardNum % 10
    if (length % 2 == 1):
        oddMul = digit * 2
        if(oddMul > 9):
            while (oddMul):
                subDigit = oddMul % 10
                oddSum += subDigit
                oddMul //= 10
        else:
            oddSum += oddMul
    else:
        evenSum += digit
    length += 1

    # Get first 2 digits
    if (cardNum < 100 and cardNum > 9):
        if (cardNum in digitsList):
            last2Digits = cardNum
        elif (cardNum in list4):
            lastDigit = 4
        else:
            valid = False
            break
    cardNum //= 10

# Check valid
if (valid == True):
    checkSum = oddSum + evenSum
    valid = ((checkSum % 10) == 0) and length > 12

# SHOW RESULT
if (valid == True and length == 15 and (last2Digits == 34 or last2Digits == 37)):
    print("AMEX")
elif (valid == True and length == 16 and (last2Digits > 50 and last2Digits < 56)):
    print("MASTERCARD")
elif (valid == True and length == 13 or length == 16 and lastDigit == 4):
    print("VISA")
elif (valid == False):
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int reversDigits(int num)
{
    int rev_num = 0;
    while(num > 0)
    {
        rev_num = rev_num*10 + num%10;
        num = num/10;
    }
    return rev_num;
}

int main(void)
{
    // PROMPT TO GET USER INPUT
    // Use get_long to check the length
    long int cardNum;
    do
    {
        cardNum = get_long("Number: ");
    }
    while (cardNum < 4);

    // REVERT NUMBER
    // int revertNum = reversDigits(cardNum);
    // printf("THIS IS REVERTED NUMBER: %i\n", revertNum);

    // GET EVERY DIGITS IN THE NUMBER
    int length = 0;
    int evenSum = 0;
    int oddMul2 = 0;
    int oddSum = 0;
    int digit;
    int checkSum;
    int lastDigit;
    int last2Digits;
    bool valid = true;
    while(cardNum)
    {
        // Get each digit
        digit = cardNum % 10;

        // Get SUM of odd digits
        if (length % 2 == 1)
        {
            // Multiply digit by 2
            oddMul2 = digit * 2;

            // If result of multiply large than9, get sum of each digits.
            if (oddMul2 > 9)
            {
                while(oddMul2)
                {
                    int subDigit = oddMul2 % 10;
                    oddSum += subDigit;
                    oddMul2 /= 10;
                }
            }
            else oddSum += oddMul2;
        }
        else
        {
            evenSum += digit;
        }
        length ++;

        // Get last 2 digits
        if (cardNum < 100 && cardNum > 9)
        {
            if (cardNum == 34 || cardNum == 37 || (cardNum > 50 && cardNum < 56))
            {
                last2Digits = cardNum;
            }

            // Get last digit
            else if (cardNum > 39 && cardNum < 50)
            {
                lastDigit = 4;
            }
            else
            {
                valid = false;
                break;
            }
        }

        cardNum /= 10;
    }

    if (valid)
    {
        // GET CHECKSUM
        checkSum = oddSum + evenSum;

        // CHECK IF '0' IS THE LAST DIGIT IN CHECK SUM
        valid = ((checkSum % 10) == 0) && length > 12;
    }

    // SHOW RESULT
    if (valid == 1 && length == 15 && (last2Digits == 34 || last2Digits == 37))
    {
        printf("AMEX\n");
    }
    else if (valid == 1 && length == 16 && (last2Digits > 50 && last2Digits < 56))
    {
        printf("MASTERCARD\n");
    }
    else if (valid == 1 && (length == 13 || length == 16) && lastDigit == 4)
    {
        printf("VISA\n");
    }
    else if (valid == 0)
    {
        printf("INVALID\n");
    }

}